File: iterator/golang/bookshelf.go
package main

// BookShelf 书架
type BookShelf struct {
	Aggragate
	books []*Book
}

// Iterator 书架迭代器
func (bs *BookShelf) Iterator() *BookShelfIterator {
	return NewBookShelfIterator(bs)
}

// NewBookShelf 初始化一个书架
func NewBookShelf() *BookShelf {
	return &BookShelf{}
}

// GetBookAt 获取指定位置的图书
func (bs *BookShelf) GetBookAt(index int) *Book {
	return bs.books[index]
}

// AppendBook 将一本图书放入书架
func (bs *BookShelf) AppendBook(b *Book)  {
	bs.books = append(bs.books, b)
}

// Len 书架上当前的图书数量
func (bs *BookShelf) Len() int {
	return len(bs.books)
}

File: iterator/php/BookShelf.php
<?php

/**
 * 书架
 */
class BookShelf implements Aggregate
{
    /**
     * @var array[Book]
     */
    private $books;
    
    /**
     * @param $index
     *
     * @return \Book
     */
    public function getBookAt($index)
    {
        return $this->books[$index];
    }
    
    /**
     * @param \Book $book
     */
    public function appendBook(Book $book)
    {
        $this->books[] = $book;
    }
    
    /**
     * @return int
     */
    public function getLength()
    {
        return count($this->books);
    }
    
    /**
     * @return \BookShelfIterator
     */
    public function iterator()
    {
        return new BookShelfIterator($this);
    }
}